src/Compilers/Core/Portable/SourceGenerator/SourceGeneratorExtensions.cs

```diff
@@ -28,25 +30,65 @@ public static class SourceGeneratorExtensions
                 throw new ArgumentNullException(nameof(path));
             }
 
-            var builder = ArrayBuilder<SyntaxTree>.GetInstance();
-            var context = new Context(builder, compilation, path, writeToDisk);
+            var treeBuilder = ArrayBuilder<SyntaxTree>.GetInstance();
+            var diagnosticBuilder = ArrayBuilder<Diagnostic>.GetInstance();
+            var context = new Context(treeBuilder, diagnosticBuilder, compilation, path, writeToDisk);
             foreach (var generator in generators)
             {
-                generator.Execute(context);
+                try
+                {
+                    generator.Execute(context);
+                }
+                catch (Exception e) when (ExceptionFilter(exceptionFilter, e))
+                {
+                    var descriptor = CreateExceptionDiagnosticDescriptor();
+                    var diagnostic = Diagnostic.Create(descriptor, Location.None, generator, e.GetType().ToString(), e.Message);
+                    diagnosticBuilder.Add(diagnostic);
+                }
+            }
+            diagnostics = diagnosticBuilder.ToImmutableAndFree();
+            return treeBuilder.ToImmutableAndFree();
+        }
+
+        private static bool ExceptionFilter(Func<Exception, bool> exceptionFilter, Exception e)
+        {
+            if (e is OperationCanceledException)
+            {
+                return false;
+            }
+
+            if (exceptionFilter == null)
```

Comment:
The suggestion isn't about correctness, it is about bypassing an invocation of a method.

---
In reply to: [68619439](https://github.com/dotnet/roslyn/pull/12205#discussion_r68619439) [](ancestors = 68619439)